controller/src/test/java/io/pravega/controller/server/eventProcessor/ScaleRequestHandlerTest.java

```diff
@@ -309,6 +311,120 @@ public void testScaleWithTransactionRequest() throws InterruptedException {
+                null, null, null, null, executor);
+        CommitRequestHandler commitRequestHandler = new CommitRequestHandler(streamStore, streamMetadataTasks, streamTransactionMetadataTasks, executor);
+
+        // 1 create transaction on old epoch and set it to committing
+        UUID txnIdOldEpoch = streamStore.generateTransactionId(scope, stream, null, executor).join();
+        VersionedTransactionData txnData = streamStore.createTransaction(scope, stream, txnIdOldEpoch, 10000, 10000,
+                null, executor).join();
+        streamStore.sealTransaction(scope, stream, txnData.getId(), true, Optional.empty(), null, executor).join();
+
+        UUID txnIdOldEpoch2 = streamStore.generateTransactionId(scope, stream, null, executor).join();
+        VersionedTransactionData txnData2 = streamStore.createTransaction(scope, stream, txnIdOldEpoch2, 10000, 10000,
+                null, executor).join();
+        streamStore.sealTransaction(scope, stream, txnData2.getId(), true, Optional.empty(), null, executor).join();
+
+        HistoryRecord epochZero = streamStore.getActiveEpoch(scope, stream, null, true, executor).join();
+        assertEquals(0, epochZero.getEpoch());
+
+        // 2. start scale
+        requestHandler.process(new ScaleOpEvent(scope, stream, Lists.newArrayList(0L),
+                Lists.newArrayList(new AbstractMap.SimpleEntry<>(0.0, 0.25), new AbstractMap.SimpleEntry<>(0.25, 0.5)), false, System.currentTimeMillis())).join();
+
+        // 3. verify that scale is complete
+        State state = streamStore.getState(scope, stream, false, null, executor).join();
+        assertEquals(State.ACTIVE, state);
+
+        // 4. just submit a new scale. dont let it run. this should create an epoch transition. state should still be active
+        streamStore.startScale(scope, stream, Lists.newArrayList(1L), Lists.newArrayList(new AbstractMap.SimpleEntry<>(0.5, 0.75), new AbstractMap.SimpleEntry<>(0.75, 1.0)),
+        System.currentTimeMillis(), false, null, executor).join();
+
+        // 5. commit on old epoch.. this should roll over.
+        assertTrue(Futures.await(commitRequestHandler.processEvent(new CommitEvent(scope, stream, txnData.getEpoch()))));
+        TxnStatus txnStatus = streamStore.transactionStatus(scope, stream, txnIdOldEpoch, null, executor).join();
+        assertEquals(TxnStatus.COMMITTED, txnStatus);
+
+        // 6. run scale. this should fail in scaleCreateNewSegments with IllegalArgumentException with epochTransitionConsistent
+        AssertExtensions.assertThrows("epoch transition should be inconsistent", requestHandler.process(new ScaleOpEvent(scope, stream, Lists.newArrayList(1L),
+                Lists.newArrayList(new AbstractMap.SimpleEntry<>(0.5, 0.75), new AbstractMap.SimpleEntry<>(0.75, 1.0)),
+                false, System.currentTimeMillis())), e -> Exceptions.unwrap(e) instanceof IllegalArgumentException);
```

Comment:
What's the reason for throwing illegal argument in this case? What's the illegal argument, the epoch transition? I feel that the argument is not really invalid, but instead inconsistent with the current state of the system.

Reply:
we can throw InputInvalidException. Would that be better?